tests/ui/invalid_pymethod_receiver.stderr

```diff
@@ -1,18 +1,32 @@
-error[E0277]: the trait bound `i32: From<&PyCell<MyClass>>` is not satisfied
+error[E0277]: `i32` is not a valid receiver type for pyo3 methods
  --> tests/ui/invalid_pymethod_receiver.rs:8:43
   |
 8 |     fn method_with_invalid_self_type(slf: i32, py: Python<'_>, index: u32) {}
-  |                                           ^^^ the trait `From<&PyCell<MyClass>>` is not implemented for `i32`
+  |                                           ^^^ the trait `pyo3::pycell::Receiver<&PyCell<MyClass>>` is not implemented for `i32`
   |
-  = help: the following other types implement trait `From<T>`:
-            <f32 as From<i16>>
-            <f32 as From<i8>>
-            <f32 as From<u16>>
-            <f32 as From<u8>>
-            <f64 as From<f32>>
-            <f64 as From<i16>>
-            <f64 as From<i32>>
-            <f64 as From<i8>>
-          and 67 others
-  = note: required because of the requirements on the impl of `Into<i32>` for `&PyCell<MyClass>`
-  = note: required because of the requirements on the impl of `TryFrom<&PyCell<MyClass>>` for `i32`
+  = note: Only `&self`, `&mut self`, `&PyCell<Self>, `&PyRef<'_, Self> and `&PyRefMut<'_, Self> are valid receiver types
+  = help: the following other types implement trait `pyo3::pycell::Receiver<T>`:
+            &'a PyCell<T>
+            PyRef<'a, T>
+            PyRefMut<'a, T>
```

Comment:
We don't need `rustc_on_unimplemented` for this, and this is already miles better. Is there a reason why we're using `TryFrom` for this rather than our own trait?

Reply:
Agreed, I think at the time it was simpler for good ol' `#[pyproto]`. I have considered making a similar change in the past, so a PR for this would be welcome!